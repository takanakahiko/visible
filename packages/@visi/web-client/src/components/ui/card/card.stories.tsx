import React from 'react';

import { Typography } from '../typography';
import { Card } from '.';

export default {
  title: 'Card',
  component: Card,
};

export const Solid = () => (
  <Card variant="solid">
    <Card.Heading>
      <Typography variant="h6">Hello world</Typography>
    </Card.Heading>
    <Card.Body>singing somewhere in the internet</Card.Body>
  </Card>
);
