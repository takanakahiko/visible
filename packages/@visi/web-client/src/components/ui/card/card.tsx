import classNames from 'classnames';
import React, { createContext, useContext } from 'react';

type CardVariant = 'solid';

const CardContext = createContext<{
  variant?: CardVariant;
}>({});

const useVariant = () => {
  return useContext(CardContext).variant;
};

export interface CardHeadingProps {
  children: React.ReactNode;
}

const Heading = (props: CardHeadingProps) => {
  const { children } = props;
  const variant = useVariant();
  const className = classNames(variant === 'solid' && ['mb-2']);

  return <header className={className}>{children}</header>;
};

export interface CardBodyProps {
  children: React.ReactNode;
}

const Body = (props: CardBodyProps) => {
  const { children } = props;
  return <div>{children}</div>;
};

export interface CardProps {
  variant: CardVariant;
  className?: string;
  children: React.ReactNode;
}

export const Card = (props: CardProps) => {
  const { variant, children, className } = props;

  return (
    <CardContext.Provider value={{ variant }}>
      <section
        className={classNames(
          'rounded',
          'p-4',
          'shadow-md',
          variant === 'solid' && ['bg-gray-300'],
          className,
        )}
      >
        {children}
      </section>
    </CardContext.Provider>
  );
};

Card.Heading = Heading;
Card.Body = Body;
