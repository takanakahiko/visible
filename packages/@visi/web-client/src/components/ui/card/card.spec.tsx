import { render } from '@testing-library/react';
import React from 'react';

import { Solid } from './card.stories';

describe('Card', () => {
  it('matches snapshot', () => {
    const { container } = render(<Solid />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
