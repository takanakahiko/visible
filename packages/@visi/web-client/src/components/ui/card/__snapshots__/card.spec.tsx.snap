// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot 1`] = `
<section
  class="rounded p-4 shadow-md bg-gray-300"
>
  <header
    class="mb-2"
  >
    <h6
      class="font-bold text-xl"
    >
      Hello world
    </h6>
  </header>
  <div>
    singing somewhere in the internet
  </div>
</section>
`;
